internal/vpmem/types.go

```diff
@@ -0,0 +1,122 @@
+package vpmem
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/Microsoft/hcsshim/internal/memory"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+const (
+	// MaxVPMEMCount is the maximum number of VPMem devices that may be added to an LCOW
+	// utility VM
+	MaxVPMEMCount = 128
+
+	// DefaultVPMemSizeBytes is the default size of a VPMem device if the create request
+	// doesn't specify.
+	DefaultVPMemSizeBytes = 4 * 1024 * 1024 * 1024 // 4GB
+
+	PageSize = 0x1000
+
+	lcowVPMemPackedLayerMountFmt  = "/run/layers/p%d-%d-%d"
+	lcowVPMemDefaultLayerMountFmt = "/run/layers/p%d"
+
+	vPMemControllerResourceFormat string = "VirtualMachine/Devices/VirtualPMem/Devices/%d"
+	vPMemDeviceResourceFormat     string = "VirtualMachine/Devices/VirtualPMem/Devices/%d/Mappings/%d"
+)
+
+var (
+	// ErrMaxVPMEMLayerSize is the error returned when the size of `hostPath` is
+	// greater than the max vPMEM layer size set at create time.
+	ErrMaxVPMEMLayerSize = errors.New("layer size is too large for VPMEM max size")
+	ErrInvalidParameter  = errors.New("invalid parameter")
+
+	ErrNoAvailableLocation = errors.New("no available location")
+	ErrNotAttached         = errors.New("not attached")
+)
+
+type deviceInfo struct {
+	hostPath     string
+	uvmPath      string
+	refCount     uint32
+	deviceNumber uint32
+}
+
+// mappedDeviceInfo is a structure used for storing information about devices mapped onto a
+// single VPMem device
+type mappedDeviceInfo struct {
+	deviceInfo
+	deviceOffset uint64
+	deviceSize   uint64
+}
+
+// virtualPMemSurface is an internal structure used for determining VPMem devices mapped to
+// a Linux utility VM
+type virtualPMemSurface struct {
+	memory.PoolAllocator
+	maxSize  uint64
+	mappings map[string]mappedVPMem
+}
+
+// genericVPMem represents a device that can be mounted onto a guest
+type genericVPMem interface {
+	// path on a host
+	HostPath() string
+	// path inside a UVM
+	MountPath() string
+	DeviceNumber() uint32
+	ResourcePath() string
+
+	// Reference counting
+	RefCount() uint32
+	AddRef() uint32
+	RemoveRef() uint32
+
+	// device settings as seen by the host
+	HostSettings() interface{}
+	// device settings as seen by the guest
+	GuestSettings() interface{}
+}
+
+// mappedVPMem represents devices that can be mounted with an offset
+type mappedVPMem interface {
+	genericVPMem
+	DeviceSize() uint64
+	DeviceOffset() uint64
+}
+
+type VirtualPMemController interface {
+	FindDevice(context.Context, string) (genericVPMem, error)
+	FindNextDevice(context.Context, string) (genericVPMem, error)
+	NewHostComputeRequest(context.Context, string, interface{}) (*hcsschema.ModifySettingRequest, error)
```

Comment:
nit: I feel like `NewModifyRequest` or `NewHostModifyRequest` would maybe be more descriptive. What do you think?

Reply:
yeah, I was going back and forth between the two lol